# Challenge 1
Below we've created a small DataFrame. You should use DataFrame API functions and UDFs to accomplish two tasks.

You need to parse the State and city into two different columns.

You need to get the number of days in between the start and end dates. You need to do this two ways.

a) Firstly, you should use SparkSQL functions to get this date difference.

b) Secondly, you should write a udf that gets the number of days between the end date and the start date.

In [ ]:
pip install Spark

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 41.0/41.0 KB 3.7 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  Created wheel for Spark: filename=spark-0.2.1-py3-none-any.whl size=58760 sha256=8d3d7630ee6eecd58afc5810ddb2df8a5350bd1bfeb2981f21bd2924470acde7
  Stored in directory: /root/.cache/pip/wheels/bc/0f/6c/b41528ca0fd4d46513185e90da91bc1c484bba6138641c5a62
Successfully built Spark


In [ ]:
pip install pyspark

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 281.4/281.4 MB 2.7 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 199.7/199.7 KB 13.7 MB/s eta 0:00:00
  Created wheel for pyspark: filename=pyspark-3.3.2-py2.py3-none-any.whl size=281824028 sha256=8e8530b68970e82e9bbe72c990626eb0f60535471403bf624749e829096fe034
  Stored in directory: /root/.cache/pip/wheels/6c/e3/9b/0525ce8a69478916513509d43693511463c6468db0de237c86
Successfully built pyspark
  Attempting uninstall: py4j
    Found existing installation: py4j 0.10.9.7
    Uninstalling py4j-0.10.9.7:
      Successfully uninstalled py4j-0.10.9.7


In [ ]:
from pyspark.sql import SparkSession
from pyspark.sql.functions import split, regexp_extract

In [ ]:
spark = SparkSession.builder.appName('test').master('local').enableHiveSupport().getOrCreate()

"id|end_date|start_date|location
1|2015-10-14 00:00:00|2015-09-14 00:00:00|CA-SF
2|2015-10-15 01:00:20|2015-08-14 00:00:00|CA-SD
3|2015-10-16 02:30:00|2015-01-14 00:00:00|NY-NY
4|2015-10-17 03:00:20|2015-02-14 00:00:00|NY-NY
5|2015-10-18 04:30:00|2014-04-14 00:00:00|CA-SD
"

In [ ]:
# cria um DataFrame com algumas informações
data = [(1,'2015-10-14 00:00:00','2015-09-14 00:00:00','CA-SF'),
        (2,'2015-10-15 01:00:20','2015-08-14 00:00:00','CA-SD'),
        (3,'2015-10-16 02:30:00','2015-01-14 00:00:00','NY-NY'),
        (4,'2015-10-17 03:00:20','2015-02-14 00:00:00','NY-NY'),
        (5,'2015-10-18 04:30:00','2014-04-14 00:00:00','CA-SD')
        ]
df = spark.createDataFrame(data, ['id','end_date','start_date','location'])

# escreve o conteúdo do DataFrame em um arquivo texto com campos separados por vírgula
df.write.csv("sample_data/fake_log0.csv", sep="|", header=True)

In [ ]:
%%sh
mv sample_data/fake_log0.csv sample_data/fake_log
mv sample_data/fake_log/fake_log0.csv sample_data/fake_log0.csv 
pwd
ls -la

total 55520
drwxr-xr-x 1 root root     4096 Mar 31 19:08 .
drwxr-xr-x 1 root root     4096 Mar 31 19:07 ..
-rwxr-xr-x 1 root root     1697 Jan  1  2000 anscombe.json
-rw-r--r-- 1 root root   301141 Mar 30 13:53 california_housing_test.csv
-rw-r--r-- 1 root root  1706430 Mar 30 13:53 california_housing_train.csv
drwxr-xr-x 2 root root     4096 Mar 31 19:10 fake_log0.csv
-rw-r--r-- 1 root root 18289443 Mar 30 13:53 mnist_test.csv
-rw-r--r-- 1 root root 36523880 Mar 30 13:53 mnist_train_small.csv
-rwxr-xr-x 1 root root      930 Jan  1  2000 README.md


mv: cannot stat 'sample_data/fake_log0.csv/*.csv': No such file or directory


In [ ]:
cat fake_log1.log

66.249.69.97 - - [24/Sep/2014:22:25:44 +0000] "GET /071300/242153 HTTP/1.1" 404 514 "-" "Mozilla/5.0 (compatible; Googlebot/2.1; +http://www.google.com/bot.html)"
71.19.157.174 - - [24/Sep/2014:22:26:12 +0000] "GET /error HTTP/1.1" 404 505 "-" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/37.0.2062.94 Safari/537.36"
71.19.157.174 - - [24/Sep/2014:22:26:12 +0000] "GET /favicon.ico HTTP/1.1" 200 1713 "-" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/37.0.2062.94 Safari/537.36"
71.19.157.174 - - [24/Sep/2014:22:26:37 +0000] "GET / HTTP/1.1" 200 18785 "-" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/37.0.2062.94 Safari/537.36"
71.19.157.174 - - [24/Sep/2014:22:26:37 +0000] "GET /jobmineimg.php?q=m HTTP/1.1" 200 222 "http://www.holdenkarau.com/" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/37.0.2062.94 Safari/537.36"

In [ ]:
%%sh
cd sample_data
ls -la 

total 55524
drwxr-xr-x 1 root root     4096 Mar 31 19:16 .
drwxr-xr-x 1 root root     4096 Mar 31 19:07 ..
-rwxr-xr-x 1 root root     1697 Jan  1  2000 anscombe.json
-rw-r--r-- 1 root root   301141 Mar 30 13:53 california_housing_test.csv
-rw-r--r-- 1 root root  1706430 Mar 30 13:53 california_housing_train.csv
drwxr-xr-x 2 root root     4096 Mar 31 19:16 fake_log
-rw-r--r-- 1 root root      272 Mar 31 19:08 fake_log0.csv
-rw-r--r-- 1 root root 18289443 Mar 30 13:53 mnist_test.csv
-rw-r--r-- 1 root root 36523880 Mar 30 13:53 mnist_train_small.csv
-rwxr-xr-x 1 root root      930 Jan  1  2000 README.md


## 1. Parse the State and city into two different columns

In [ ]:
df = spark.read.format("csv").options(header='true', delimiter = '|').load("sample_data/fake_log0.csv")

In [ ]:
df.createOrReplaceTempView('tpv_df')

In [ ]:
spark.sql('''
select location
      ,substr(location,1,instr(location,"-") -1) as State
      ,substr(location,instr(location,'-')+1,length(location) ) as City
from tpv_df
''').show(truncate=False)

+--------+-----+----+
|location|State|City|
+--------+-----+----+
|CA-SF   |CA   |SF  |
|CA-SD   |CA   |SD  |
|NY-NY   |NY   |NY  |
|NY-NY   |NY   |NY  |
|CA-SD   |CA   |SD  |
+--------+-----+----+



# Challenge 2

The following is log data represents data colelcted from a web server application.

## 1. Parsing Logs
Parse the logs in to a DataFrame/Spark SQL table that can be queried.

Schema to use is as follow:

ip: String = 66.249.69.97

date: String = 24/Sep/2014

methodType: String = GET

uri: String = /071300/242153

protocol: String = HTTP/1.1

response: String = 404

size: String = 514

agent: String = Mozilla/5.0 (compatible; Googlebot/2.1; +http://www.google.com/bot.html)

In [ ]:
# Lendo o arquivo em um DataFrame
log_lines = spark.read.csv("fake_log1.log",inferSchema=True)
log_lines.show(truncate=False)

+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
|_c0                                                                                                                                                                               |
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
|66.249.69.97 - - [24/Sep/2014:22:25:44 +0000] "GET /071300/242153 HTTP/1.1" 404 514 "-" "Mozilla/5.0 (compatible; Googlebot/2.1; +http://www.google.com/bot.html)"                |
|71.19.157.174 - - [24/Sep/2014:22:26:12 +0000] "GET /error HTTP/1.1" 404 505 "-" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML                                       |
|71.19.157.174 - - [24/Sep/2014:22:26:12 +0000] "GET /favicon.ico HTTP/1.1" 200 1713 "-" "Mozil

In [ ]:
# criando TempView
log_lines.createOrReplaceTempView('tpv_log')

+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
|_c0                                                                                                                                                                               |
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
|66.249.69.97 - - [24/Sep/2014:22:25:44 +0000] "GET /071300/242153 HTTP/1.1" 404 514 "-" "Mozilla/5.0 (compatible; Googlebot/2.1; +http://www.google.com/bot.html)"                |
|71.19.157.174 - - [24/Sep/2014:22:26:12 +0000] "GET /error HTTP/1.1" 404 505 "-" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML                                       |
|71.19.157.174 - - [24/Sep/2014:22:26:12 +0000] "GET /favicon.ico HTTP/1.1" 200 1713 "-" "Mozil

In [ ]:
spark.sql("""
select substr(_c0,1,instr(_c0,"-") -2) as IP
      ,substr(_c0, instr(_c0,"[")+1,11) as date
      ,instr(_c0,'"') as tt
      ,split(substr(_c0,instr(_c0,'"'))," ") as full_text
      ,_c0
from tpv_log
""").show(truncate=False)

+-------------+-----------+---+------------------------------------------------------------------------------------------------------------------------------------------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
|IP           |date       |tt |full_text                                                                                                                                       |_c0                                                                                                                                                                               |
+-------------+-----------+---+------------------------------------------------------------------------------------------------------------------------------------------------+----------------------------------------------------------------------------------------------------------------

In [ ]:
# exibindo dados
spark.sql('''
select IP
      ,Date
      ,replace(full_text[0],'"','') as Method_Type 
      ,full_text[1] as uri
      ,replace(full_text[2],'"','') as Protocol
      ,full_text[3] as Response
      ,full_text[4] as Size
      ,replace(full_text[6],'"','') as Agent
from (
select substr(_c0,1,instr(_c0,"-") -2) as IP
      ,substr(_c0, instr(_c0,"[")+1,11) as date
      ,instr(_c0,'"') as tt
      ,split(substr(_c0,instr(_c0,'"'))," ") as full_text
      ,_c0
from tpv_log) log
''').show(truncate=False)

+-------------+-----------+-----------+-------------------+--------+--------+-----+-----------+
|IP           |Date       |Method_Type|uri                |Protocol|Response|Size |Agent      |
+-------------+-----------+-----------+-------------------+--------+--------+-----+-----------+
|66.249.69.97 |24/Sep/2014|GET        |/071300/242153     |HTTP/1.1|404     |514  |Mozilla/5.0|
|71.19.157.174|24/Sep/2014|GET        |/error             |HTTP/1.1|404     |505  |Mozilla/5.0|
|71.19.157.174|24/Sep/2014|GET        |/favicon.ico       |HTTP/1.1|200     |1713 |Mozilla/5.0|
|71.19.157.174|24/Sep/2014|GET        |/                  |HTTP/1.1|200     |18785|Mozilla/5.0|
|71.19.157.174|24/Sep/2014|GET        |/jobmineimg.php?q=m|HTTP/1.1|200     |222  |Mozilla/5.0|
+-------------+-----------+-----------+-------------------+--------+--------+-----+-----------+



## 2. Analysis

Generate some insights fromg the log data, e.g.: Get the most visited pages.

Sorry guys, my profile is Data Engineer **not** Data Analyst or Data Scientist. I guess you need to count the total of "response" from each "uri" value found. 
If the highest value founded in a "uri" with "responde" value was "404", you have a problem probably. "404" regarding http codes means "not found".

# Challenge 3

Build an SCD Type 2 table based on the following data

___

Conteúdo dos arquivos

File 1 [
{"message_id": 1, "name": "Johnsie Nunez", "email": "civic1907@example.org", "date_of_birth": "2009-01-15", "city": "Durham"},
{"message_id": 2, "name": "Hong Hays", "email": "spokesman1981@example.org", "date_of_birth": "2019-01-25", "city": "Oxford"},
{"message_id": 3, "name": "Broderick Flynn", "email": "tales1891@example.org", "date_of_birth": "2001-02-06", "city": "Hastings"},
{"message_id": 4, "name": "Virgil Hooper", "email": "boundaries2076@example.org", "date_of_birth": "2009-12-27", "city": "Nottingham"},
{"message_id": 5, "name": "Delmar Richmond", "email": "crawford1926@example.org", "date_of_birth": "2005-11-07", "city": "Durham"},
{"message_id": 6, "name": "Elin Hahn", "email": "folder1831@example.org", "date_of_birth": "2001-08-02", "city": "Liverpool"},
{"message_id": 7, "name": "Latrina Kent", "email": "blank1874@example.org", "date_of_birth": "2002-05-01", "city": "Cambridge"}
]


File 2
[
{"message_id": 1, "name": "Michael Scott", "email": "mscott@example.org", "date_of_birth": "2011-01-15", "city": "Durham"},
{"message_id": 8, "name": "Peter Parker", "email": "thespider123@example.org", "date_of_birth": "2007-01-21", "city": "Liverpool"},
]

Specifications

Use message_id as the identifier of data changes

Use date_of_birth (asc order) as the value to expire data changes

Latest data changes (that are not expired) should be Null for experiaton date

In [ ]:
%%sql
CREATE TABLE if not exists my_schema.tbl_contacts
(id     integer null
,name   nvarchar(max) null
,email  nvarchar(max) null
,date_of_birth  date null
,city   nvarchar(max) null
,date_creation  date 
,date_end       date null
,is_active      boolean
);
CREATE NONCLUSTERED INDEX idx_tbl_contacts_id (
    ON my_schema.tbl_contacts (id));